xapian-core/tests/harness/backendmanager.cc

```diff
@@ -145,14 +155,36 @@ BackendManager::get_database(const std::string &dbname,
     tmp_path += '~';
 
     {
-	Xapian::WritableDatabase wdb = get_writable_database(tmp_dbleaf,
-							     string());
+	Xapian::WritableDatabase wdb;
+	if (is_remote_bm) {
+	    wdb = get_writable_database_from_sub_manager(tmp_dbleaf);
+	} else if (is_honey_bm) {
+	    wdb = get_alt_writable_database(tmp_dbleaf, string());
+	} else {
+	    wdb = get_writable_database(tmp_dbleaf, string());
+	}
 	gen(wdb, arg);
     }
     rename(tmp_path.c_str(), path.c_str());
     // For multi, the shards will use the temporary name, but that's not really
     // a problem.
 
+    if (is_remote_bm) {
+	return get_remote_database_by_name(dbleaf, 30000);
+    }
+    if (is_honey_bm) {
+	// converting a glass backend to honey
+	string tmpfile = path;
+	tmpfile += ".tmp";
+
+	Xapian::WritableDatabase wdb(path);
+	wdb.compact(tmpfile, Xapian::DB_BACKEND_HONEY);
+	wdb.close();
+
+	rm_rf(path);
+
+	rename(tmpfile.c_str(), path.c_str());
+    }
     return Xapian::Database(path);
```

Comment:
... and then a `finalise_generated_database()` virtual method could implement this part.

Or have a single method which does both and the `gen()` and `rename()` steps in between get duplicated.

Reply:
I tried fixing it a little. I'll fix the code style issues once you're satisfied with the current changes.

Reply:
I think the structure is looking better, but there's still some dispatching based on checking the name (`is_remote_bm`).

You've already added a `get_remote_database_by_name()` virtual method, but with a default implementation which throws an exception.  If you instead made the default implementation `return Xapian::Database(path);` (and
added a parameter to pass in `path`) then we wouldn't need `is_remote_bm` here at all (`finalise_generated_database()` does nothing by default, and the remote managers don't override that).  That lets the virtual method do the dispatch for us.

The method name would make rather less sense then - I'd suggest renaming to `get_database_by_name()`.

I'd probably not pass in the timeout from here.  We have to hard-code it somewhere in the backend manager machinery for this case, and doing so in the remote tcp manager at least means less backend-specific stuff here.  

Also the timeouts used for other remote tcp databases in the test harness are all `300000` (5 minutes) not `30000` (30 seconds).  Probably either is long enough to avoid timeouts during runs even under heavy load in practice, but the timeout used by default really should be consistent.

Reply:
Added `get_database_by_path` function, and removed `get_remote_database_by_name`.

Reply:
I've made `finalise_generated_database` take a name instead of a path. If I modify it(basically make `get_database_by_name` a replacement of `finalise_generated_database`) I will have to perform messy string operations inside honey's backendmanager. So I've not removed `finalise_generated_database()`.